【提出物】コメント機能追加課題 by ポリモーフィック

SeijiNumata opened the pull request.

SeijiNumata committed:
コメント機能を実装しました。

コメントが投稿できるようにしました。

SeijiNumata committed:
reportのCLUDを実装

scafolldを利用して、一旦report機能を実装しました。

SeijiNumata committed:
タイムゾーンの変更

投稿日時を日本時間で表示するため、タイムゾーンを変更しました

SeijiNumata committed:
N+1問題に対応

コメント一覧ページにincludeを追加して、N+1問題に対応

SeijiNumata committed:
コメント欄にpagingの対応

kaminariでコメント欄をpagingできるようにしました。

SeijiNumata committed:
作業途中だったi18nを作成、全て編集済み

作業途中のi18nを全て完了させました。

SeijiNumata committed:
不要ファイル、スペースを削除、インデントを整える

不必要なものを削除し、フォーマットを整えました。

SeijiNumata committed:
コメント詳細ページが見れないバグを修正

過去の余計なコメントを削除しました。

SeijiNumata committed:
Rubocopの修正（フォーマットを綺麗にしました)

一箇所不要なスペースがあるとRubocopに指摘されたので直しました。

SeijiNumata committed:
日報の詳細ページの編集ボタンが、作成者以外は見れないようにviewを修正

上記の通りです。元から編集はできませんでしたが、ボタンが見えてしまっていたので修正しました。

betachelsea commented on app/views/books/show.html.erb:
```diff
@@ -22,3 +22,4 @@
 
 <%= link_to t('views.common.edit'), edit_book_path(@book) %> |
 <%= link_to t('views.common.back'), books_path %>
+<%= render 'comments/comment', commentable: @book,comment: @comment,comments: @comments %> 
```
この `@comment`, `@comments` は元を辿ると `@book` (commentable) から取得できるものなので、うまくやると引数が減らせそうかなと思いました :eyes:  うーん、この辺の実装は好みもあるかなとは思うのですが、おそらく `commentable` と `comment`、 `comments` はこのケースでは不可分なので、全部 commentable からもらっちゃっても問題ないかなと感じます。

SeijiNumata replied:
> 全部 commentable からもらっちゃっても問題ないかなと感じます。

すみません、こちら難しくてできなかったです😭
確かに引数が多くて名前が似ていて分かりにくいのでなんとかしたいというのはあるのですが、詳しく教えていただければ嬉しいです。
***

まず、整理します。

コントローラから渡される
`@comment`は`@comment = Comment.new`です。
`@comments`は`@comments = @book.comments.includes(:user).order(:id).page(params[:page])` です。

このそれぞれをcommentableからもらうということですが、`@comment`については全然やり方が分かりません。
`@comments`は、`commentable.comments.includes(:user).order(:id).page(params[:page])`と書けばcommentableからもらえますが、これをview側に記述するという認識で良いのでしょうか？

お手数おかけして申し訳ありません。
よろしくお願いします。🙇‍♂️

betachelsea replied:
> @commentsは、commentable.comments.includes(:user).order(:id).page(params[:page])と書けばcommentableからもらえますが、これをview側に記述する

はい、そのつもりで書きました 🙇 
ただ `params[:page]` とかをpartial内で呼び出すことになるので、ちょっとややこしくなるかなとは思い、人によっては逆にviewに書くほうが煩雑に見えてよく無い、という意見もありそうかなと... 💦 
seijiさんがどうしたいかで決めていただいて大丈夫です！ 🙏 

> @commentについて

すみません、こちら改めて考えたら、showアクションとcreateアクションのタイミングで保持すべきインスタンス内容が違うので、controller側で `@comment` を定義して呼び出す、のほうがやっぱりスマートそうかなと改めて読んで思いました!! すみません!! 🙇‍♂️ 
今回のケースでは、view側でinputのrequired指定により、ブラウザ側でbodyが存在しないことを警告してくれます。また、Commentモデルでもvalidationの指定が特にないので、controller側にエラーかどうかを確認するケースが事実上ありません。ですが、今後Railsを使っていくと、controller側でエラーを検知した場合、入力予定だったデータを保持してviewに渡すというやり方が必要になってきます。

```ruby
# show アクション
# 中身が空っぽ
@comment = Comment.new

# create アクション
# ユーザが入力しようとした値を保持している
@comment = @commentable.comments.build(comment_params)
```

うっかりブラウザ側の挙動で問題なさそうだったので Comment.new をpartial内呼び出したらスッキリしそう...と思っちゃったのですが、今後のことを考えるとNGな指摘でした。申し訳ないです！ 🙏 
`@comment` はcontrollerで定義したものを渡していきましょう。

SeijiNumata replied:
> @commentsのcommentable.comments.includes(:user).order(:id).page(params[:page]) を、viewに書くかcontrollerに書くかの問題

なるほど！好みって感じですね！
引数が増える問題がありますが、ページネーションの部分は他のviewと統一してコントローラーから貰った方が分かりやすい　という理由から、このままにしておきたいと思います！

そのように実装することもあると知れて良かったです！
今度から他のpartialで引数が増えてしまった時に、対応できそうです。
ありがとうございます！！
***

> `@comment` について 

> showアクションとcreateアクションのタイミングで保持すべきインスタンス内容が違うので、controller側で `@comment` を定義して呼び出す、のほうがやっぱりスマートそうかなと改めて読んで思いました!!

なるほど〜〜！
その時のご指摘の意図と、「`@comment`をなぜコントローラーで呼ぶべきか」ということに対して、良く理解できました！確かにそうですね、ご丁寧にありがとうございます！

自分はここまで深く考えられていなかったので、とても納得感がありました。

ありがとうございます。

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,71 @@
+  # GET /reports/new
+  def new
+    @report = Report.new
+    @current_user = current_user
```
`current_user` はdeviseのヘルパーメソッドで、viewからも呼び出せるので、 `@current_user` には入れなくても大丈夫そうです 👀

SeijiNumata replied:
その通りですね！
修正しました。

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,26 @@
+end
+
+def set_commentable
+  resource, id = request.path.split('/')[1, 2]
```
この辺の実装の出典はこのブログですかね？ 結構勢いのある実装ですね！ 👀 
[ポリモーフィック関連のコントローラー \- 猫Rails](http://nekorails.hatenablog.com/entry/2019/06/13/031003)

> request情報から強引に取得します。
> ちなみにこちらの方法を使ってコントローラーを分けると、より綺麗な実装になります。

ブログにある通り、このやり方だとroutes.rb (path) の内容に実装が依存していて、結構強引なやり方だなと感じました。一階層増やすとかで、routes.rbを変更したら簡単に壊れちゃうなと。
"こちら" である動画の15分あたりから、このへんの実装をControllerの継承で解決する方法が紹介されています。
https://gorails.com/episodes/comments-with-polymorphic-associations


上記が少し難しいと感じる場合は、pathに含まれるパラメータ類は `params[:report_id]` といった表現で取得することができるので、その内容の有無で対象リソースを判断するのもありかなと思いました。やりはじめるとif文が増殖しちゃいそうとは思うのですが、routes.rbの修正によって壊れることはあまりない状態にできそうです。
参考: https://railsguides.jp/routing.html

```sh
$ rails routes | grep comment

book_comments POST     /books/:book_id/comments(.:format)     comments#create
report_comments POST     /reports/:report_id/comments(.:format)   comments#create                                                        

# これらの :book_id 、 :report_id の部分が params[:book_id]、 params[:report_id] として取得できます。
```

...という感じで、やり方は最終的にはおまかせするのですが、`request.path.split` を使わない方法への修正チャレンジをしてみてもらえますでしょうか！ 🙏

SeijiNumata replied:
> この辺の実装の出典はこのブログですかね？ 結構勢いのある実装ですね！ 👀
ポリモーフィック関連のコントローラー - 猫Rails

そのブログを参考に実装しました！
私も、この強引な実装方法ちょっと嫌だな〜と思っていたんです。
ただ、技術力と検索力がなく、これ以上どうしたらいいか分かりませんでした。

突っ込んでもらえて、また具体的な対応策も示していただいて本当に嬉しいです。ありがとうございます。

動画を見てControllerの継承で解決する方法を試してみたら上手くできました。
`params[:report_id]`でも取れることを確認しましたが、今回は折角なので継承の実装で提出いたします。

このようにコントローラで共通して使いたいものもあるけど、処理が変わる時もある！という時に、継承を使えばいいんですね〜〜〜。
本で読んで理解をしていても、実際に使えませんでした。

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,71 @@
+  end
+
+  def contribute_user?
+    return true if current_user == @report.user
```
controllerでチェックすると、viewを開発者ツールとかでいじられて不正にリクエストが飛んできても、ちゃんと防げてよいですね！ 👍

SeijiNumata replied:
皆さんの日報を拝見してやってみました！
ありがとうございます。👍

betachelsea commented on app/views/comments/_comment.html.erb:
```diff
@@ -0,0 +1,34 @@
+    <% comments.each do |comment|%>
+    <tr>
+      <td> <%= comment.body %></td>
+      <% if comment.user.name=="" %>
```
```suggestion
      <% if comment.user.name.blank? %>
```
こんな感じだとRailsっぽいですね〜

SeijiNumata replied:
確かにそうですね〜
この辺の慣れていないところは、１つずつ直していきたいと思います。

betachelsea commented on app/views/comments/_comment.html.erb:
```diff
@@ -0,0 +1,34 @@
+<h2><%= t('comments.post') %></h2>
+
+<%= form_with model: [commentable,comment ] do |f| %>
+<%= f.text_area :body ,:required => true %>
+<%= f.hidden_field :user_id,  :value => current_user.id %><br>
+<%= f.submit  %>
+<% end %>
```
```suggestion
<%= form_with model: [commentable, comment] do |f| %>
    <%= f.text_area :body, :required => true %>
    <%= f.hidden_field :user_id, :value => current_user.id %><br>
    <%= f.submit %>
<% end %>
```
ものすごく細かいところなのですが、実際の現場では(とても急いでいたら見過ごされるかもしれないけど...)できるだけ綺麗なコードが喜ばれると思っています。半角スペース開ける、詰める、インデント...といったところは都度意識してやっていくと、nヶ月後に見直した時も読みやすく、修正しやすい、メンテが楽なコードになっていくと思います！
他の箇所ももしあれば意識してもらえれば嬉しいです！

SeijiNumata replied:
丁寧に見ていただいてありがとうございます！！

erbの自動整形ってないのでしょうかね・・・
htmlやCSSは自動整形がありますが、erbはhtmlの自動整形を代わりに使っても、細かいところが修正できないようです。(vscodeのbeautifyという拡張機能を使っています)

こちら手動で修正いたしました。
もし自動整形が存在しなければ、これからも手動で直していく必要がありそうですね。

betachelsea replied:
自分はiTerm2でvimを使ってゴリゴリ書いちゃっててvscode未履修ですね... 💦 
やはりerbとhtmlは別物なので ( `<%= %>` とかはhtmlに無いので... )、htmlの自動整形をerbに適用してもうまくいかなそうかと。

ちなみに、beautify では settings.json でerbも指定できるようですが、こちらすでにご利用中でしょうか？念の為共有します! 🙏 
[【簡単】vsCodeでRubyのerbファイルを自動フォーマットする方法 \- Qiita](https://qiita.com/Kiyo_Karl2/items/6fc6e0b57512deb1ff60)

SeijiNumata replied:
> ちなみに、beautify では settings.json でerbも指定できるようですが、こちらすでにご利用中でしょうか？念の為共有します! 🙏
【簡単】vsCodeでRubyのerbファイルを自動フォーマットする方法 - Qiita

共有ありがとうございます！

この設定は反映していいて、ある程度は直してくれるのですが、それでも`<%= %>`の中などは直してもらえないんです😭
全てをAutoFormatに頼るという訳には行かなそうなので、細かいところは自分の手で直していきたいと思います！

betachelsea commented on app/views/reports/index.html.erb:
```diff
@@ -0,0 +1,30 @@
+        <td><%= report.title %></td>
+        <td><%= report.description %></td>
+        <td><%= link_to t('views.common.show'), report %></td>
+        <% if current_user == report.user %>
```
`current_user == report.user` 、何回か登場してきている気がしますね 👀 
こういう時は report モデルに、current_user (に限らず、userモデルのインスタンス)を引数としてとってくる判定メソッドを生やしておくと、将来もし編集権限を与えるユーザの条件に仕様変更( ...例えば、ユーザに課金レベルが出来て課金している人しか編集できないようにしてください、とか... )があったとき、何箇所も変更をいれずに済んで便利です。

メソッド化に挑戦してみてください〜 🙏

SeijiNumata replied:
確かにDRYにした方がいいですね〜　メソッド化してみました！👍

一点質問があります。
メソッドをヘルパーに書くか、モデルに書くか　というのは何を基準に決めるのでしょうか？
https://qiita.com/Yukaaaah/items/19e524fd0c0e4a3451f1
をみると、「viewで使うものはヘルパーで定義しよう！」　ということを書いていますが、ネットの記事はあまり参考にならないことも多いので・・・
他にも調べてみましたが、viewはヘルパー、controllerはモデルでメソッドを定義する　みたいな基準のものが多かったです。

自分の意見では、モデルメソッドにはモデルが関係するもの全てが入るのかな？と思いました。
だとしたら、ここの処理は後々report以外でも使うという要件(bookで同じことをしたい、新規に映画を追加したいと言った場合)にも対応できるのでヘルパーで定義しても良いかなと思いました。

よく分からないので、今回は両方で実装して、両方で使ってみました。
(背景を知らない人が見たら分かりにくいので、本来はどちらか片方のみ実装して使うべきだと思います。)

こちらご意見を伺えると幸いです。
よろしくお願いします...!!

betachelsea replied:
自分は大まかに次のように考えています。 💭 

- ヘルパー
    - 基本的にViewのみで頻出する機能をメソッド化する(Controllerで使いたい時はincludeしなきゃいけなくなるので、Viewでの問題解決に注力したい)
    - 複数のモデルを跨いで考えなければいけないことを解決させる
- モデル
    - view以外で頻出する機能をメソッド化する
    - 定義されたモデル以外のデータに手を出さないようにする

両方で実装できるケースは悩ましいな...と思いますが、view以外での出番があり、また明確に特定のモデルの機能であると名言できそうな場合は、モデルに定義したほうが使いやすそうだなと考えます。

ここでおまけ知識となりますが、Railsに最初から入ってはいないのですが、decorator層という考え方を紹介しておきます。基本的にviewで使いたいけど、modelに書くと肥大化するし、helperにやらせるにしてもちょっと細かすぎてしんどい、みたいなときに導入されると便利です。例えば

```ruby
user.created_at.strftime('%Y年%m月%d日')
```

みたいなことはviewに出す上でよくあるケースだと思うのですが、こういう系がずらっとModelに並んでも大変なので、それを取りまとめておく場所としてdecoratorという概念があります。
詳しくは調べて見てもらうと良さそうです〜 ※今回別に使わなくても大丈夫です！将来出番があれば。

[decoratorを導入して、viewの記述をすっきりさせ、modelの肥大化を回避する【Day 3/30 2nd】｜K\.Tawara｜note](https://note.com/kentarotawara/n/n0c1ff2060cf7)

まぁ何事もほどほど、という話もありますが... 😅 
[Railsアンチパターン: Decoratorの肥大化（翻訳）｜TechRacho（テックラッチョ）〜エンジニアの「？」を「！」に〜｜BPS株式会社](https://techracho.bpsinc.jp/hachi8833/2018_03_06/53535)


いずれも、リポジトリ毎に取り扱う中身は違うので、絶対これ！という正解がない話になってしまうと思いますが、自分の考え方としてご紹介しました 🙏   
将来実際にチームでやるときは、チームでのレビューの中で相談して決めていくと思われます！ seijiさんもどういうやり方が一番良いか、引き続き模索してみてもらえればと思います 🚀

SeijiNumata replied:
おお!! めちゃめちゃ参考になります！！🤩

これは私にとって、ヘルパーとモデルのどっちにメソッドを書くか問題の、１つの指針になりそうです。
decoratorという概念も知らなかったので、すごく嬉しいです！

どういうやり方が一番良いか、自分の中で１つ柱を持っていると、実装に迷いがなくなりますし、チームで議論できるときも意見できるので良さそうですね！

ありがとうございます！！

betachelsea commented on app/views/reports/index.html.erb:
```diff
@@ -0,0 +1,30 @@
+        <td><%= link_to t('views.common.show'), report %></td>
+        <% if current_user == report.user %>
+        <td><%= link_to t('views.common.edit'), edit_report_path(report) %></td>
+        <td><%= link_to t('views.common.destroy'), report, method: :delete, data: { confirm: 'Are you sure?' } %></td>
```
他の箇所がi18n適用済みなので、ここのconfirm文もやっておきましょうか。

SeijiNumata replied:
気がつくことができずに、抜けてしまいました。

修正いたしました。

betachelsea commented on config/credentials.yml.enc:
```diff
@@ -1 +0,0 @@
-/81Xb3f6e24YKrW8OkBPQzb57XVSD7HjtVwW7VZqc9Q04pgisboNZop/AHW0DBEEchzDK9QqiQ0O5NS8x+Birjxi4t0mHH13kCuS2OMquqMpBpllH/F4HseHoudhOQCiK+0xfxDbwra59YFn5d8DZjPcp5BeWVyJAfTyg/CZxXOYvl1wBzb9zOC/qLEsiJDv9iw3xxq4WaSEfD7Dqed5lu88W5YDpepjLz1hiIrsIhB6nLTDbBgiGBq26mh+5GxB8U5E1lYKjA9Ea0jyaB+BaP1cAMwcwkvOLZXx2L5bz6GFl12tbyxES/U4Soq+6zrIiAkZHocBnx6GLaEJuAwKq/FDR7i/07a3vUHSiyNzvVfHAP7E6VU9fJpBeTE7ZbFe2RQnhOlcuUuzr9p ...
```
あれ、credentials.yml.enc いらなかったですか？ 🤔  (あまり消すことのないファイルだなと思いまして)

***
すみません、なぜかコメントに返信が送れないので編集で回答します。

過去に関係のないmaster.keyを作ってしまったため、どちらかのファイルを消さないといけなくなってしまったのです。
これはその名残りなのですが、今回プラクティスに関係ないということで、毎回メンターさんに突っ込まれるのですが見逃してもらっています。

すみません、よろしくお願いします。

betachelsea replied:
なるほど！失礼しました！ 🙆  🙇

betachelsea commented on app/views/reports/index.html.erb:
```diff
@@ -0,0 +1,30 @@
+<p id="notice"><%= notice %></p>
```
どうやら、 noticeが二重に出てしまっているようですね。なぜこうなったのか？を確認し、解決をお願いします！

![スクリーンショット 2021-02-04 6 57 28](https://user-images.githubusercontent.com/2603449/106816598-f4ae2080-66b8-11eb-807b-44eedfc122bc.png)

SeijiNumata replied:
再現確認しました。

原因ですが、
reports/index.html.erbと、application_html.erbで`flash[:notice]`が２回呼ばれていました。
reports/index.html.erbの方で呼んでいる`flash[:notice]`は不要なため、削除いたしました。

また、こちらのバリデーションのflashは、noticeよりalertの方が良いと判断しましたので、noticeからalertに変更いたしました。

betachelsea commented on app/views/reports/_form.html.erb:
```diff
@@ -0,0 +1,31 @@
+  </div>
+
+  <div class="field">
+    <%= form.hidden_field :user_id,  :value => current_user.id %>
```
current_user を紐付けたい場合、viewではなくcontroller側で指定したほうがより安全です。
一番簡単にできる不正例としては、例えば開発者ツールとかでここのvalueの値をいじってから更新すると、他ユーザになりすまして投稿することができます 📝 
current_userをcontroller側で設定するように変更してみてもらえますでしょうか？ 👀

SeijiNumata replied:
なるほど、、、それは危険ですね。
どうやってコントローラ側でcurrent_userを登録するんだろう？と思ったのですが、`report_params`に`.merge`というメソッドがあることを知り、実装してみました。

一般的なやり方かどうか分からないので、もしそうでなければ「普通はこうする」というのを教えていただきたいです。

betachelsea replied:
`merge` の利用で良いと思います！　対応ありがとうございます 👍 
別の案としては、次のようにするケースもあります。

```ruby
@report = Report.new(report_params)
@report.user = current_user
```

どちらをえらぶかは...また好みかもしれないです！ 😅 
既存リポジトリを修正するときは、既存コード内で採用されているやり方に揃えるようにすると、一貫性が生まれて良さそうです。

SeijiNumata replied:
問題ない様で良かったです！
```ruby
@report = Report.new(report_params)
@report.user = current_user
```

この書き方、思いつかなかったです！
ありがとうございます！！

betachelsea reviewed (commented):
コメントしました！Railsプラクティスの終盤ということもあり、ちょっと細かめに拝見しました 🙏 👀  不明な点ありましたら追加でコメントしていただければと思います！

SeijiNumata committed:
レビュー反映：set_commentableの実装が力づくだったので、継承で実装するように変更

上記ままです。　comments_contollerでbooksとreports分の対応をしていましたが、処理が異なるので継承で実装するようにしました。

SeijiNumata committed:
レビュー反映：current_userをインスタンス変数にわざわざしていたので削除

deviseのヘルパーメソッドなので、必要ありませんでした。

SeijiNumata committed:
レビュー反映: _comment.html.erbに渡していた引数を３つから１つに変更

commentable,comment,commentsを渡していたが、３つ渡す必要はなかったため１つに変更

SeijiNumata committed:
レビュー反映　フラッシュが重複している不具合の対応

上記の通りです。片方消しました。

SeijiNumata committed:
レビュー対応　current_user=report.user　というのをメソッド化

ヘルパーに書くのか、モデルに書くのかという問題があるので、質問してみます。

SeijiNumata committed:
i18n抜けしてしまったものがあったので修正

上記の通りです。

SeijiNumata committed:
indentがずれてしまっているものを修正

上記の通りです。コードに修正はありません。

SeijiNumata committed:
rubocopをかけました。

rubocopをかけました。内容の修正はありません。

SeijiNumata committed:
before_actionが継承先でダブっていたので、片方削除

SeijiNumata committed:
=="" とやっていたところを、.blank?とRailsっぽい書き方を意識した書き方に変更

上記の通りです。

SeijiNumata committed:
コメント投稿がうまくいかなくなってしまっていたので、戻してうまくいく状態で一時保存

パーシャルに引数を渡す関係だと思うのですが、戻せるようにコミットしておきます。

SeijiNumata committed:
user_idを渡していなかったため、バリデーションに引っかかってしまっていたと判明したため修正

hidden.fieldで隠してuseridを渡していましたが、セキュリティ上の観点からコントローラで渡すように修正
 Please enter the commit message for your changes. Lines starting

SeijiNumata committed:
bookのコメント投稿がうまく行っていなかったので修正

さらに、パーシャルに渡す引数を１つにしたいという実装をここからします。

SeijiNumata committed:
ファイルの末尾に改行がないものがあったため修正

上記の通りです

SeijiNumata reviewed (commented):
ありがとうございます！！😭
細かく見ていただいた方が勉強になるので嬉しいです！

修正いたしました。
コメントのパーシャルに渡す引数のところで、分からないところがあったので、コメントしています。

ご確認よろしくお願いします。🙇‍♂️

betachelsea commented:
@SeijiNumata  修正ありがとうございました！
今回のプラクティスの学習のポイントとして、Controllerの継承があったかなと思うのですが、その部分を修正していただきましたので確認OKとします 🎉  
残りは結構好みによる部分もあったり、自動修正が効いてないせいだったりとのことで、もうちょっと直すかどうするかはseijiさんにお任せしたいと思います。
Commentのpartialの件も、意見変更しちゃってすみません!! 🙇 

もし気になるところあったらもちろん追加でみますので、リプライいただければと思います。よろしくお願いします！

SeijiNumata commented:
@betachelsea 
ご丁寧にレビュー、解説していただいてありがとうございます！！😁

確認OKとのことで、私としてもやり切りましたので追加の対応はせずにこれで完了したいと思います!!
Commentのpartialの件も、スッキリしたので良かったです！色々考えてくださりありがとうございます。

質問等も今のところありませんので大丈夫そうです!!
また何かあればよろしくお願いいたします。
ありがとうございます！🙇‍♂️
